README
Spiff WikiMarkup
-----------------
This library is part of the Spiff platform.

Spiff WikiMarkup is a library for (bidirectionally) converting between Wiki
markup language and HTML.


Contact
--------
Mailing List: http://groups.google.com/group/spiff-devel/


Dependencies
-------------
None


Usage
------
FIXME (email the author if you need docs, that will make him write it)


ChangeLog
2009-01-08 Samuel Abels <http://debain.org>

	* Release 0.4.0.

Changes since 0.3

	* Html2Wiki: Support for XHTML span attributes with class "bold", "italic",
	and "underline" was added.

	* Wiki2Html: Underline using XHTML strict compliant <span class="underline">
	instead of <u> tags.

Changes since 0.2

	* Html2Wiki.py: A broken unit test caused many bugs to remain undiscovered.
	The test was fixed, the linebreak-handling of the class was reworked and some
	other bugs were also fixed. You should particulary notice a much higher
	reliability when converting between wiki markup and HTML multiple times.
	Prepending a "!" before a WikiWord now prevents it from being linked.
	URL captions now allow more characters.
	Links are now made by enclosing them in square brackets (no longer curly
	brackets).
	Support for preformatted text with the #Text/#End tag was added.
	Improved recognition of external URLs.

Changes since 0.1

	* Wiki2Html.py: Now also works with windows line breaks. Line break handling
	has also bee improved in other cases. The _underline_ tag now underlines, it
	no longer strikes through the text.
	The #Table tag has been removed, tables work now only by using the #Heading
	and #Row tags.
	Headings now support additional characters, like punctuation, slashes,
	dashes, etc.
	(set_wiki_word_handler): New method, provides a hook into the new WikiWord
	parser.
	(set_url_handler): New method, provides a hook into the new URL parser.

	* Html2Wiki.py: Added support for hyperlinks. Changed to reflect the changes
	in Wiki2Html.py.


tests/SpiffWikiMarkup/CellTest.py
import sys, unittest, re, os.path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', 'src'))

def suite():
    tests = ['testCell']
    return unittest.TestSuite(map(CellTest, tests))

from SpiffWikiMarkup.Cell import Cell

class CellTest(unittest.TestCase):
    def testCell(self):
        cell = Cell()

if __name__ == '__main__':
    unittest.TextTestRunner(verbosity = 2).run(suite())


tests/SpiffWikiMarkup/run_suite.py
#!/usr/bin/python
import os, sys, unittest
modules = ['CellTest',
           'Html2WikiTest',
           'WikiParserTest',
           'Wiki2HtmlTest']

# Parse CLI options.
if len(sys.argv) == 1:
    verbosity = 2
elif len(sys.argv) == 2:
    verbosity = int(sys.argv[1])
else:
    print 'Syntax:', sys.argv[0], '[verbosity]'
    print 'Default verbosity is 2'
    sys.exit(1)

# Load all test suites.
all_suites = []
for name in modules:
    module = __import__(name, globals(), locals(), [''])
    all_suites.append(module.suite())

# Run.
suite = unittest.TestSuite(all_suites)
unittest.TextTestRunner(verbosity = verbosity).run(suite)


tests/SpiffWikiMarkup/Html2WikiTest.py
import sys, unittest, re, os.path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', 'src'))

def suite():
    tests = ['testHtml2Wiki', 'testConvert']
    return unittest.TestSuite(map(Html2WikiTest, tests))

import os
from SpiffWikiMarkup import Html2Wiki, Wiki2Html

class Html2WikiTest(unittest.TestCase):
    def testHtml2Wiki(self):
        # Read the entire file into one string.
        filename = 'markup.txt'
        infile   = open(filename, 'r')
        in_str   = infile.read()
        infile.close()

        # Parse the file.
        parser = Html2Wiki()
        parser.feed('''<b>test</b>\n''')

        #print '"%s"' % parser.wiki
        self.assert_(parser.wiki == '''*test*''')


    def testConvert(self):
        # Read the entire file into one string.
        filename = 'markup.txt'
        infile   = open(filename, 'r')
        in_str   = infile.read()
        infile.close()

        # Convert Wiki to HTML.
        parser = Wiki2Html()
        parser.read(filename)
        html1 = parser.html
        #print html1

        # Convert the HTML back to Wiki.
        parser = Html2Wiki()
        parser.feed(html1)
        wiki = parser.wiki
        #print wiki

        # Write the wiki to a file.
        # The result contains an extra newline to mark a table end.
        # This is generally ok, but would break the unit test, so we
        # remove the last character before writing.
        fd = open(filename + '.tmp', 'w')
        fd.write(wiki[:-1])
        fd.close()

        # Convert the new Wiki file to HTML again.
        parser = Wiki2Html()
        parser.read(filename + '.tmp')
        html2 = parser.html
        #print html2

        # Clean up.
        os.remove(filename + '.tmp')

        # Make sure that the model is complete.
        self.assert_(len(in_str) > 10)
        self.assert_(len(html1)  > 10)
        self.assert_(html1 == html2)

if __name__ == '__main__':
    unittest.TextTestRunner(verbosity = 2).run(suite())


src/SpiffWikiMarkup/Html2Wiki.py
# Copyright (C) 2006 Samuel Abels, http://debain.org
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2, as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
import HTMLParser, re, sys
from Cell import Cell

class Html2Wiki(HTMLParser.HTMLParser):
    def __init__(self):
        HTMLParser.HTMLParser.__init__(self)
        self.wiki       = ''
        self.buffer     = ''
        self.indent     = 0
        self.linebreak  = '\n'
        self.rows       = []
        self.cells      = []
        self.in_table   = False
        self.in_td      = False
        self.in_heading = False
        self.in_ul      = False
        self.in_ol      = False
        self.in_li      = False
        self.in_a       = False
        self.in_pre     = False
        self.last_href  = ''
        self.span_path  = []

    def __output(self, text, linebreak = True):
        self.buffer += (' ' * self.indent * 2)
        self.buffer += text + self.linebreak
        
    def __flush(self):
        self.wiki  += self.buffer
        self.buffer = ''
        
    def handle_starttag(self, tag, attrs):
        if   tag == 'table':  self.start_table()
        elif tag == 'tr':     self.start_tr()
        elif tag == 'th':     self.start_th(attrs)
        elif tag == 'td':     self.start_td(attrs)
        elif tag == 'h1':     self.start_h1()
        elif tag == 'h2':     self.start_h2()
        elif tag == 'h3':     self.start_h3()
        elif tag == 'ul':     self.start_ul()
        elif tag == 'ol':     self.start_ol()
        elif tag == 'li':     self.start_li()
        elif tag == 'i':      self.start_i()
        elif tag == 'b':      self.start_b()
        elif tag == 'u':      self.start_u()
        elif tag == 'a':      self.start_a(attrs)
        elif tag == 'pre':    self.start_pre()
        elif tag == 'strike': self.start_strike()
        elif tag == 'span':   self.start_span(attrs)
        elif tag == 'br':     self.newline()
        
    def handle_endtag(self, tag):
        if   tag == 'table':  self.end_table();
        elif tag == 'tr':     self.end_tr()
        elif tag == 'th':     self.end_th()
        elif tag == 'td':     self.end_td()
        elif tag == 'h1':     self.end_h1()
        elif tag == 'h2':     self.end_h2()
        elif tag == 'h3':     self.end_h3()
        elif tag == 'ul':     self.end_ul()
        elif tag == 'ol':     self.end_ol()
        elif tag == 'li':     self.end_li()
        elif tag == 'i':      self.end_i()
        elif tag == 'b':      self.end_b()
        elif tag == 'u':      self.end_u()
        elif tag == 'a':      self.end_a()
        elif tag == 'pre':    self.end_pre()
        elif tag == 'strike': self.end_strike()
        elif tag == 'span':   self.end_span()

    def start_h1(self):
        self.buffer += '='

    def end_h1(self):
        self.buffer += '=\n\n'

    def start_h2(self):
        self.buffer += '=='

    def end_h2(self):
        self.buffer += '==\n\n'

    def start_h3(self):
        self.buffer += '==='

    def end_h3(self):
        self.buffer += '===\n\n'

    def start_ul(self):
        self.in_ul = True

    def end_ul(self):
        self.in_ul = False

    def start_ol(self):
        self.in_ol = True

    def end_ol(self):
        self.in_ol = False

    def start_li(self):
        self.in_li = True
        if self.in_ol:
            self.buffer += '# '
        elif self.in_ul:
            self.buffer += '* '

    def end_li(self):
        self.in_li = False

    def start_i(self):
        self.buffer += '/'

    def end_i(self):
        self.buffer += '/'

    def start_b(self):
        self.buffer += '*'

    def end_b(self):
        self.buffer += '*'

    def start_u(self):
        self.buffer += '_'

    def end_u(self):
        self.buffer += '_'

    def start_a(self, attrs):
        self.in_a = True
        self.last_href = ''
        for key, value in attrs:
            if key == 'href':
                self.last_href = value
        self.buffer += '[' + self.last_href

    def end_a(self):
        self.in_a = False
        self.buffer += ']'

    def start_pre(self):
        self.in_pre = True
        self.buffer += '#Text\n'

    def end_pre(self):
        self.in_pre = False
        self.buffer += '#End\n'

    def start_strike(self):
        self.buffer += '-'

    def end_strike(self):
        self.buffer += '-'

    def start_span(self, attrs):
        cls = None
        for key, value in attrs:
            if key == 'class':
                cls = value
        if cls == 'underline':
            char = '_'
        elif cls == 'bold':
            char = '*'
        elif cls == 'italic':
            char = '/'
        else:
            char = ''
        self.buffer += char
        self.span_path.append(char)

    def end_span(self):
        self.buffer += self.span_path.pop()

    def start_table(self):
        self.in_table = True
        
    def start_tr(self):
        pass

    def start_th(self, attrs):
        self.in_heading = True
        self.start_td(attrs)

    def start_td(self, attrs):
        self.__flush()
        self.in_td = True
        cell       = Cell()
        for key, value in attrs:
            if key == 'rowspan':
                cell.rowspan = int(value)
            elif key == 'colspan':
                cell.colspan = int(value)
        self.cells.append(cell)

    def handle_data(self, data):
        if not self.in_pre:
            data = data.replace('\n', '')
        if self.in_a:
            if data == self.last_href:
                return
            self.buffer += ' '
        if self.in_li:
            self.buffer += data.strip() + '\n'
        if self.in_ul or self.in_ol:
            self.__flush()
        elif self.in_td:
            self.buffer += data
        elif not self.in_table:
            self.buffer += data
            self.__flush()

    def end_td(self):
        self.cells[-1].data += self.buffer.strip()
        self.buffer = ''
        self.in_td = False

    def end_th(self):
        self.end_td()

    def end_tr(self):
        if len(self.cells) is 0:
            return
        if self.in_heading:
            self.__output('#Heading')
            self.in_heading = False
        else:
            self.__output('#Row')
        self.indent += 1
        line = ('|' * self.cells[0].colspan) + ' ' + self.cells[0].data.strip()
        for cell in self.cells[1:]:
            line += ' ' + ('|' * cell.colspan) + ' ' + cell.data.strip()
        if len(line) <= 80:
            self.__output(line)
        else:
            for cell in self.cells:
                self.__output(('|' * cell.colspan) + ' ' + cell.data.strip())
        self.cells   = []
        self.indent -= 1
        self.__flush()

    def end_table(self):
        self.in_table = False
        self.__flush()

    def newline(self):
        self.buffer += '\n'


src/SpiffWikiMarkup/Wiki2Html.py
# Copyright (C) 2006 Samuel Abels, http://debain.org
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2, as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
import HTMLParser, re, sys
from WikiParser import WikiParser
from cgi        import escape

class Wiki2Html(object):
    def __init__(self):
        self.html              = ''
        self.buffer            = ''
        self.counter           = 0
        self.in_list           = False
        self.in_numbered_list  = False
        self.in_heading        = False
        self.in_table          = False
        self.in_cell           = False
        self.indent_level      = 0
        self.url_handler       = None
        self.wiki_word_handler = None


    def set_url_handler(self, handler):
        """
        Whenever the parser finds a URL in the output, by default it
        automatically translates it into a clickable link in the HTML output.
        However, if you want more fine-grained control over which URLs are
        converted into links, you might want to use this function.
        It calls the given handler function before a link is added into the
        HTML. The function is passed the URL and the caption.
        The handler can the modify the URL and the caption, and pass it
        back to the parser. Example handler:
        
        def url_handler(url, caption):
            return (url, 'My Caption')
        
        The parser uses the returned values to build a link in the HTML.
        If URL is None, the item is not linked at all.
        """
        self.url_handler = handler


    def set_wiki_word_handler(self, handler):
        """
        Like set_url_handler() but for WikiWords.
        Example handler:
        
        def wiki_word_handler(word):
            return (url, word)
        
        The parser uses the returned values to build a link in the HTML.
        If URL is None, the item is not linked at all.
        """
        self.wiki_word_handler = handler


    def indent(self, level):
        self.indent_level = level


    def dedent(self, level):
        self.indent_level = level


    def code_start(self, text):
        self.buffer += '<pre>'


    def code(self, text):
        self.buffer += escape(text)


    def code_end(self, text):
        self.buffer += '</pre>\n'


    def title1_start(self, text):
        self.buffer += '<h1>'


    def title1_end(self, text):
        self.buffer += '</h1>'


    def title2_start(self, text):
        self.buffer += '<h2>'


    def title2_end(self, text):
        self.buffer += '</h2>'


    def title3_start(self, text):
        self.buffer += '<h3>'


    def title3_end(self, text):
        self.buffer += '</h3>'


    def italic_start(self, text):
        self.buffer += '<i>'


    def italic_end(self, text):
        self.buffer += '</i>'


    def bold_start(self, text):
        self.buffer += '<b>'


    def bold_end(self, text):
        self.buffer += '</b>'


    def underline_start(self, text):
        self.buffer += '<span class="underline">'


    def underline_end(self, text):
        self.buffer += '</span>'


    def unnumbered_list_start(self, text):
        self.buffer += '<ul>\n'
        self.in_list = True


    def numbered_list_start(self, text):
        self.buffer += '<ol>\n'
        self.in_list = True
        self.in_numbered_list = True


    def list_end(self, text):
        if self.in_numbered_list:
            self.buffer += '</ol>\n'
            self.in_numbered_list = False
        else:
            self.buffer += '</ul>\n'
        self.in_list = False


    def list_item_start(self, text):
        self.buffer += '<li>'


    def list_item_end(self, text):
        self.__flush(True)
        self.buffer += '</li>\n'


    def table_start(self, text):
        self.in_table = True
        self.counter  = 0
        self.buffer += '<table>\n'


    def heading_start(self, text):
        self.in_heading = True
        self.row_start(text)


    def row_start(self, text):
        self.buffer  += '<tr>'


    def cell_start(self, text):
        self.in_cell = True
        attribs = ''
        if len(text) - 1 > 1:
            attribs += ' colspan="%i"' % (len(text) - 1)
        if self.in_heading:
            self.buffer += '<th%s>' % attribs
        else:
            self.buffer += '<td%s>' % attribs
        self.__flush(True)


    def cell_end(self, text):
        # If the cell contains only a hash, replace it by a counter.
        if self.buffer.strip() == '#':
            self.counter += 1
            self.buffer = str(self.counter) + '.'
        self.__flush(True)
        self.in_cell = False
        if self.in_heading:
            self.html += '</th>'
        else:
            self.html += '</td>'
        self.html += '\n' + (' ' * self.indent_level)


    def row_end(self, text):
        self.buffer += '</tr>\n' + (' ' * self.indent_level)


    def heading_end(self, text):
        self.in_heading = False
        self.row_end(text)


    def table_end(self, text):
        self.in_table = False
        self.buffer += '</table>\n'


    def text(self, text):
        self.buffer += escape(text).replace('\n', '<br/>\n')


    def newline(self, text):
        if (self.in_table
            or self.in_list
            or (self.buffer[-5:-2] == '</h' and self.buffer[-1] == '>')
            or (self.buffer[-6:-3] == '</h' and self.buffer[-2] == '>')):
            self.buffer += '\n'
            return
        self.buffer += '<br/>\n'


    def link(self, text):
        text = text[1:-1]
        pos = text.find(' ')
        if pos == -1:
            url     = text
            caption = url
        else:
            url     = text[:pos]
            caption = text[pos + 1:]
        if self.url_handler is not None:
            (url, caption) = self.url_handler(url, caption)
        url     = escape(url)
        caption = escape(caption)
        if url is None:
            self.buffer += caption
        else:
            self.buffer += '<a href="%s">%s</a>' % (url, caption)


    def wiki_word(self, text):
        if text.startswith('->'):
            text = text[2:]
        url     = text
        caption = text
        if self.wiki_word_handler is not None:
            (url, caption) = self.wiki_word_handler(url, caption)
        url     = escape(url)
        caption = escape(caption)
        if url is None:
            self.buffer += caption
        else:
            self.buffer += '<a href="%s">%s</a>' % (url, caption)


    def not_wiki_word(self, text):
        self.buffer += escape(text)


    def __flush(self, strip = False):
        if strip:
            self.html += self.buffer.strip()
        else:
            self.html += self.buffer
        self.buffer = ''


    def read(self, filename):
        infile    = open(filename, 'U')
        parser    = WikiParser(infile, filename)
        nonecount = 0
        self.html = ''
        while True:
            token    = parser.read()
            position = parser.position()
            if token[0] is None:
                nonecount += 1  # This is because Plex is broken.
            if nonecount >= 2:
                break
            try:
                method = getattr(self, token[0])
            except:
                method = None
            if method is not None:
                method(token[1])
        infile.close()
        self.__flush()


src/SpiffWikiMarkup/WikiParser.py
# Copyright (C) 2006 Samuel Abels, http://debain.org
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2, as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
import sys, string
from Plex import *

# Single char definitions.
letter      = Range('AZaz')
digit       = Range('09')
spaces      = Any(" \t\r\n")
nl          = Str("\n") | Eof
not_nl      = AnyBut("\r\n")
dash        = Str('-')
colon       = Str(':')
hash        = Str('#')
at          = Str('@')
equal       = Str('=')
star        = Str('*')
slash       = Str('/')
o_bracket   = Str('{')
c_bracket   = Str('}')
underscore  = Str('_')
dot         = Str('.')
punctuation = Any('.!?,;:')
quotes      = Any('\'"')

# Single word definitions.
name        = letter + Rep(letter | digit)
file_name   = Rep1(letter | digit | dash | underscore | dot)
wiki_force  = Str('->') + name
wiki_word   = Range('AZ') + name + Range('AZ') + name
not_w_word  = Str('!') + wiki_word
indentation = Rep(Str(' ')) | Rep(Str("\t"))
variable    = letter + Rep(letter | digit | Str('_')) + Rep(letter | digit)
proto       = Alt(Str('http'), Str('https'), Str('ftp'), Str('mailto'))
login       = file_name + Opt(colon + file_name)
host        = name + Rep(letter | digit | dash | underscore | dot) + name
path        = Opt(slash) + file_name + Rep(slash + file_name) + Opt(slash)
arg_value   = Rep1(letter | digit | dash | underscore | dot | Str('%'))
arg         = file_name + equal + Opt(arg_value)
args        = Str('?') + arg + Rep(Str('&') + arg)
path_args   = path + Opt(args)
url         = proto + colon + Str('//') + Opt(login + at) + host + Opt(path_args)
code        = Str('#Text') + nl
code_end    = Str('#End') + nl

# Markup.
line          = Rep(not_nl) + nl
blank_line    = indentation + nl
words         = Rep1(letter | digit | spaces)
list_item     = Bol + Alt(hash, star) + Str(' ')
italic_start  = Alt(Bol, spaces) + slash
italic_end    = slash + Alt(Eol, spaces, punctuation)
phrase        = Rep1(words | dash | at | punctuation | slash | quotes| hash)
title1        = equal + phrase + equal
title2        = equal + equal + phrase + equal + equal
title3        = equal + equal + equal + phrase + equal + equal + equal
heading       = Str('#Heading') + nl
row           = Str('#Row') + nl
cell          = Rep1(Str('|')) + Str(' ')
internal_link = Str('[') + url + Opt(spaces + phrase) + Str(']')
external_link = Str('[') + path_args + Opt(spaces + phrase) + Str(']')

class WikiParser(Scanner):
    def __init__(self, file, filename):
        Scanner.__init__(self, self.lexicon, file, filename)
        self.my_buffer             = ''
        self.indentation_stack     = [0]
        self.bracket_nesting_level = 0
        self.in_italic             = False
        self.in_bold               = False
        self.in_underline          = False
        self.in_list               = False
        self.in_list_item          = False
        self.in_table              = False
        self.in_heading            = False
        self.in_row                = False
        self.in_cell               = False
        self.begin('indent')

    def _buffer_flush(self):
        indent = self._get_current_indent_level()
        #print "_buffer_flush (%i): '%s'" % (indent, self.my_buffer)
        if self.my_buffer == '':
            return
        if self.my_buffer == '\n':
            self.produce('newline', '\n')
        else:
            self.produce('text', self.my_buffer)
        #print "Flush produced"
        self.my_buffer = ''

    def _newline_action(self, text):
        #print '_newline_action'
        self._buffer_flush()
        self.produce('newline', text)
        if self.bracket_nesting_level == 0:
            self.begin('indent')

    def _indentation_action(self, text):
        #print '_indentation_action'
        self.my_buffer += text
        self._buffer_flush()
        current_level = self._get_current_indent_level()
        new_level     = len(text)
        if new_level > current_level:
            self._indent_to(new_level)
        elif new_level < current_level:
            self._dedent_to(new_level)
        self.begin('')

    def _indent_to(self, new_level):
        self.indentation_stack.append(new_level)
        self.produce('indent', new_level)

    def _dedent_to(self, new_level):
        while new_level < self._get_current_indent_level():
            self.indentation_stack.pop()
        self.produce('dedent', self._get_current_indent_level())

    def _get_current_indent_level(self):
        return self.indentation_stack[-1]

    def _code_start(self, text):
        #print '_code_start: ' + text
        self._buffer_flush()
        self.produce('code_start', text)
        self.begin('code')

    def _code(self, text):
        #print '_code: ' + text
        self.my_buffer += text

    def _code_end(self, text):
        #print '_code_end: ' + text
        if self.my_buffer != '':
            self.produce('code', self.my_buffer)
            self.my_buffer = ''
        self.produce('code_end', text)
        self.begin('')

    def _title1(self, text):
        #print '_title1: ' + text
        self._buffer_flush()
        self.produce('title1_start', text[0])
        self.my_buffer += text[1:-1]
        self._buffer_flush()
        self.produce('title1_end', text[-1])

    def _title2(self, text):
        #print '_title2: ' + text
        self._buffer_flush()
        self.produce('title2_start', text[0:2])
        self.my_buffer += text[2:-2]
        self._buffer_flush()
        self.produce('title2_end', text[-2:])

    def _title3(self, text):
        #print '_title3: ' + text
        self._buffer_flush()
        self.produce('title3_start', text[0:3])
        self.my_buffer += text[3:-3]
        self._buffer_flush()
        self.produce('title3_end', text[-3:])

    def _italic_start(self, text):
        #print '_italic_start "%s"' % text
        if text[0] != '/':
            self.my_buffer += text[0]
        self._buffer_flush()
        self.in_italic = True
        self.produce('italic_start', text[-1])

    def _italic_end(self, text):
        #print '_italic_end'
        self._buffer_flush()
        self.produce('italic_end', text[0])
        self.in_italic = False
        self.my_buffer += text[1]

    def _bold(self, text):
        #print '_bold'
        self._buffer_flush()
        if self.in_bold:
            self.produce('bold_end', text)
            self.in_bold = False
        else:
            self.in_bold = True
            self.produce('bold_start', text)

    def _underline(self, text):
        #print '_underline'
        self._buffer_flush()
        if self.in_underline:
            self.produce('underline_end', text)
            self.in_underline = False
        else:
            self.in_underline = True
            self.produce('underline_start', text)

    def _list_item(self, text):
        self._buffer_flush()
        if not self.in_list:
            if text.startswith('#'):
                self.produce('numbered_list_start', '')
            elif text.startswith('*'):
                self.produce('unnumbered_list_start', '')
            self.in_list = True
        if self.in_list_item:
            self._list_item_end()
        self.in_list_item = True
        self.produce('list_item_start')
    
    def _list_item_end(self):
        self.in_list_item = False
        self.produce('list_item_end', '')

    def _list_end(self):
        if self.in_list_item:
            self._list_item_end()
        self.in_list = False
        self.produce('list_end', '')

    def _heading_start(self, text):
        self._buffer_flush()
        if not self.in_table:
            self.produce('table_start', '')
            self.in_table = True
        if self.in_heading:
            self._heading_end()
        self.in_heading = True
        self.produce('heading_start', text[:-1])
        self._newline_action(text[-1])

    def _row_start(self, text):
        self._buffer_flush()
        if not self.in_table:
            self.produce('table_start', '')
            self.in_table = True
        if self.in_heading:
            self._heading_end()
        if self.in_row:
            self._row_end()
        self.in_row = True
        self.produce('row_start', text[:-1])
        self._newline_action(text[-1])

    def _cell_start(self, text):
        self._buffer_flush()
        if not self.in_heading and not self.in_row:
            self._text(text)
            return
        if self.in_cell:
            self._cell_end()
        self.in_cell = True
        self.produce('cell_start')

    def _cell_end(self):
        self.in_cell = False
        self.produce('cell_end', '')

    def _row_end(self):
        if self.in_cell:
            self._cell_end()
        self.in_row = False
        self.produce('row_end', '')

    def _heading_end(self):
        if self.in_cell:
            self._cell_end()
        self.in_heading = False
        self.produce('heading_end', '')

    def _table_end(self):
        if self.in_heading:
            self._heading_end()
        if self.in_row:
            self._row_end()
        self.in_table = False
        self.produce('table_end', '')

    def _link(self, text):
        self._buffer_flush()
        self.produce('link', text)

    def _wiki_word(self, text):
        self._buffer_flush()
        self.produce('wiki_word', text)

    def _not_wiki_word(self, text):
        self._buffer_flush()
        self.produce('not_wiki_word_intro', text[0])
        self.produce('not_wiki_word', text[1:])

    def _open_bracket_action(self, text):
        #print '_open_bracket_action'
        self.my_buffer += text
        self.bracket_nesting_level = self.bracket_nesting_level + 1

    def _close_bracket_action(self, text):
        #print '_close_bracket_action'
        self.my_buffer += text
        self.bracket_nesting_level = self.bracket_nesting_level - 1

    def _blank_line(self, text):
        self._dedent_to(0)
        self.my_buffer += text
        if self.in_list:
            self._list_end()
        if self.in_table:
            self._table_end()
        self._buffer_flush()

    def _text(self, text):
        #print "Char: '%s'" % text
        self.my_buffer += text

    def eof(self):
        #print "EOF"
        self._blank_line('')

    lexicon = Lexicon([
        # Handle whitespace and indentation.
        (nl, _newline_action),
        State('indent', [
            (blank_line,  _blank_line),
            (indentation, _indentation_action)
        ]),

        # Preformatted text/code.
        (code,     _code_start),
        State('code', [
            (code_end, _code_end),
            (AnyChar,  _code)
        ]),

        # Brackets.
        (o_bracket, _open_bracket_action),
        (c_bracket, _close_bracket_action),

        # Styles.
        (list_item,    _list_item),
        (title1,       _title1),
        (title2,       _title2),
        (title3,       _title3),
        (star,         _bold),
        (underscore,   _underline),
        (italic_start, _italic_start),
        (italic_end,   _italic_end),

        # Tables.
        (heading, _heading_start),
        (row,     _row_start),
        (cell,    _cell_start),

        # Other.
        (internal_link, _link),
        (external_link, _link),
        (not_w_word,    _not_wiki_word),
        (wiki_word,     _wiki_word),
        (wiki_force,    _wiki_word),
        (AnyChar,       _text)
    ])
